### PR TITLE
Add one to many associations for box and location

### LINKED ISSUE #384: Add new association (assembly_location) to Box
We'll need to know the Location that the Box is being Assembled in, so we can find the correct InventoryTallies to increment and decrement. This issue is setting up the back end to handle this data.

- Add to `Box`: `belongs_to :assembly_location ...`
- Add to `Location`: `has_many :assembled_boxes ...`
- Add `assembly_location` dropdown to Box form that produces a list of Locations with type 'storage_unit'
- Add migration to connect Box to Location `add_reference :boxes, :assembly_location ...`
(look up named associations in Rails to complete the above steps)
https://edgeguides.rubyonrails.org/association_basics.html#options-for-belongs-to-foreign-key
https://www.theodinproject.com/courses/ruby-on-rails/lessons/active-record-associations

### PR DESCRIPTION
**Pull request by mauricio-prl**
Resolves https://github.com/rubyforgood/voices-of-consent/issues/384

### Description
This PR is adding one to many associations for box and location, and adding [shoulda matchers](https://github.com/thoughtbot/shoulda-matchers) to test it.

### Type of change

- New feature

### How Has This Been Tested?
Checking if the tables have the associations with rspec and shoulda matchers.

### CODE REVIEW

File: spec/rails_helper.rb [hunk 1]
```diff
@@ -38,6 +38,13 @@
   exit 1
 end
 RSpec.configure do |config|
+  Shoulda::Matchers.configure do |config|
```
Comment (andrerpbts) [inline spec/rails_helper.rb]: I'd suggest you to move this code to `spec/support/shoulda.rb`

File: app/views/boxes/_form.html.erb [hunk 2]
```diff
@@ -10,6 +10,8 @@
     <%= f.association :shipping_payment %>
     <%= f.input :shipped_at %>
     <%= f.input :shipment_tracking_number %>
+    <%= f.select :assembly_location, Location.pluck(:name, :id),
```
Comment (andrerpbts) [inline app/views/boxes/_form.html.erb]: Maybe ordering here by location names would turn the user life easier to find the desired record...

Comment (mauricio-prl) [reply app/views/boxes/_form.html.erb]: The sort method is the better way?

File: spec/rails_helper.rb (hunk 1 above)
Comment (mauricio-prl) [reply spec/rails_helper.rb]: done :+1:

File: app/views/boxes/_form.html.erb [hunk 3]
```diff
@@ -10,6 +10,8 @@
     <%= f.association :shipping_payment %>
     <%= f.input :shipped_at %>
     <%= f.input :shipment_tracking_number %>
+    <%= f.select :assembly_location, Location.pluck(:name, :id).sort,
```
Comment (andrerpbts) [inline app/views/boxes/_form.html.erb]: Avoid doing queries in the view. I suggest you to change this to:

```ruby
@assembly_locations = Location.order(:name).pluck(:name, :id) # in controller (available in new, edit, create, update)
```

Then, just add it here...

```erb
<%= f.select :assembly_location, @assembly_locations, ...
```

Comment (mauricio-prl) [reply app/views/boxes/_form.html.erb]: Thanks! it's done :+1:

File: spec/models/box_spec.rb [hunk 4]
```diff
@@ -326,4 +326,6 @@
       )
     end
   end
-end
\ No newline at end of file
+
+  it { is_expected.to belong_to(:assembly_location).class_name('Location').optional }
```
Comment (brunohkbx) [inline spec/models/box_spec.rb]: Shoulda also supports to check the `inverse_of` attribute.

File: spec/models/location_spec.rb [hunk 5]
```diff
@@ -34,4 +34,6 @@
     location.destroy!
     expect(volunteer.reload.university_location).to be_nil
   end
+
+  it { is_expected.to have_many(:assembled_boxes).class_name('Box') }
```
Comment (brunohkbx) [inline spec/models/location_spec.rb]: Here you could also check for `foreign_key` and `inverse_of` attributes.

File: spec/models/box_spec.rb (hunk 4 above)
Comment (mauricio-prl) [reply spec/models/box_spec.rb]: Nice! thanks!

File: spec/models/location_spec.rb (hunk 5 above)
Comment (mauricio-prl) [reply spec/models/location_spec.rb]: :+1: Thanks!

**Review by maebeale — APPROVED** [review]
@m-pereira hello and thank you for this PR!!! we won't know the assembly location of the box until the BoxAssembly phase. i'm going to approve this PR, but make a new issue that allows for a Box to be created without an associated `assembly_location`.

**Pull request merged**